File: deposit_wallets_ts/8_market_sell_order.ts
/**
 * 8_market_sell_order.ts — FOK market SELL YES tokens at best bid - 1 tick.
 *
 * Run: npx tsx 8_market_sell_order.ts
 * Must have bought YES tokens first (7_market_buy_order.ts).
 */

import "dotenv/config";
import { AssetType, Side, OrderType } from "@polymarket/clob-client-v2";
import { createPublicClient, http, getAddress } from "viem";
import { polygon } from "viem/chains";
import { makeClient, DEPOSIT_WALLET, RPC_URL, validSize } from "./_clob.js";

const COND_ID      = "0x7ad403c3508f8e3912940fd1a913f227591145ca0614074208e0b962d5fcc422";
const YES_TOKEN_ID = "16040015440196279900485035793550429453516625694844857319147506590755961451627";

const CTF = getAddress("0x4D97DCd97eC945f40cF65F87097ACe5EA0476045");
const ERC1155_ABI = [{
    name: "balanceOf", type: "function", stateMutability: "view",
    inputs: [{ name: "account", type: "address" }, { name: "id", type: "uint256" }],
    outputs: [{ name: "", type: "uint256" }],
}] as const;

const client = await makeClient();

// Sync YES token balance to CLOB
await client.updateBalanceAllowance({ asset_type: AssetType.CONDITIONAL, token_id: YES_TOKEN_ID });

const publicClient = createPublicClient({ chain: polygon, transport: http(RPC_URL) });
const rawBal   = await publicClient.readContract({ address: CTF, abi: ERC1155_ABI, functionName: "balanceOf", args: [getAddress(DEPOSIT_WALLET), BigInt(YES_TOKEN_ID)] });
const holdings = Number(rawBal) / 1e6;
if (holdings < 0.01) { console.error("⚠ No YES tokens — run 7_market_buy_order.ts first"); process.exit(1); }

const info     = await client.getClobMarketInfo(COND_ID);
const tickSize = String(info.mts ?? "0.01") as "0.1" | "0.01" | "0.001" | "0.0001";
const tick     = parseFloat(tickSize);
const negRisk  = info.nr ?? false;

const book = await client.getOrderBook(YES_TOKEN_ID);
const bids = ((book as any).bids ?? []).sort((a: any, b: any) => parseFloat(b.price) - parseFloat(a.price));
if (!bids.length) { console.error("No bids — cannot market sell"); process.exit(1); }

// 1 tick below best bid to guarantee fill
const price  = Math.round((parseFloat(bids[0].price) - tick) * 10000) / 10000;
// For market SELL: amount = shares to sell
const amount = validSize(holdings, price);
if (amount <= 0) { console.error(`⚠ No valid size for price=${price}`); process.exit(1); }

console.log(`YES holdings : ${holdings.toFixed(5)} tokens`);
console.log(`Placing FOK market SELL: ${amount} shares @ $${price}  (best bid: $${bids[0].price})`);

const resp = await client.createAndPostMarketOrder(
    { tokenID: YES_TOKEN_ID, price, amount, side: Side.SELL },
    { tickSize, negRisk },
    OrderType.FOK,
);
const orderId = (resp as any).orderID;
const status  = (resp as any).status;
console.log(`\n✓ Order result : ${orderId}  status: ${status}`);

File: deposit_wallets/8_market_sell_order.py
"""
8_market_sell_order.py — FOK market SELL YES tokens at best bid - 1 tick (deposit wallet).

Run:
    python 8_market_sell_order.py

Requires: PRIVATE_KEY, DEPOSIT_WALLET, CLOB_V2_BASE_URL, POLYGON_RPC_URL in .env
Sells YES tokens from the deposit wallet at market. Must have bought them first (7_market_buy_order.py).
signatureType=3 (POLY_1271).
"""

import os
import sys
from math import gcd
sys.path.insert(0, os.path.dirname(__file__))

from dotenv import load_dotenv
from web3 import Web3
from py_clob_client_v2 import (
    AssetType,
    BalanceAllowanceParams,
    ClobClient,
    OrderArgs,
    OrderType,
    PartialCreateOrderOptions,
    Side,
    SignatureTypeV2,
)

load_dotenv(dotenv_path=os.path.join(os.path.dirname(__file__), ".env"))

COND_ID        = "0x7ad403c3508f8e3912940fd1a913f227591145ca0614074208e0b962d5fcc422"  # JD Vance 2028
YES_TOKEN_ID   = "16040015440196279900485035793550429453516625694844857319147506590755961451627"

PRIVATE_KEY    = os.environ["PRIVATE_KEY"]
DEPOSIT_WALLET = os.environ["DEPOSIT_WALLET"]
CLOB_HOST      = os.environ.get("CLOB_V2_BASE_URL", "https://clob.polymarket.com")
RPC_URL        = os.environ.get("POLYGON_RPC_URL", "https://polygon-bor-rpc.publicnode.com")

CTF = Web3.to_checksum_address("0x4D97DCd97eC945f40cF65F87097ACe5EA0476045")
ERC1155_ABI = [
    {"name": "balanceOf", "type": "function", "stateMutability": "view",
     "inputs": [{"name": "account", "type": "address"}, {"name": "id", "type": "uint256"}],
     "outputs": [{"name": "", "type": "uint256"}]},
]


def fok_size(holdings: float, price: float) -> float:
    """
    Largest valid FOK sell size ≤ holdings where:
      - size has ≤5 decimal places        (maker/shares constraint)
      - size * price has ≤2 decimal places (taker/pUSD constraint)

    For price p/q the valid step = (q*1000 / gcd(p, q*1000)) / 1e5
    """
    s = f"{price:.10f}".rstrip("0")
    d = len(s.split(".")[1]) if "." in s else 0
    q = 10 ** d
    p = round(price * q)
    mod  = q * 1000
    step = (mod // gcd(p, mod)) / 1e5
    n    = int(holdings / step)
    return round(n * step, 5)


w3 = Web3(Web3.HTTPProvider(RPC_URL))
DW = Web3.to_checksum_address(DEPOSIT_WALLET)

temp_client = ClobClient(CLOB_HOST, chain_id=137, key=PRIVATE_KEY,
                         signature_type=SignatureTypeV2.POLY_1271, funder=DEPOSIT_WALLET)
creds  = temp_client.create_or_derive_api_key()
client = ClobClient(CLOB_HOST, chain_id=137, key=PRIVATE_KEY, creds=creds,
                    signature_type=SignatureTypeV2.POLY_1271, funder=DEPOSIT_WALLET)
print(f"Deposit wallet : {DW}  |  API key: {creds.api_key} ✓\n")

info      = client.get_clob_market_info(COND_ID)
token_id  = YES_TOKEN_ID
tick_size = str(info.get("mts", "0.01"))
tick      = float(tick_size)
neg_risk  = info.get("nr", False)

# Sync YES token balance to CLOB
client.update_balance_allowance(
    BalanceAllowanceParams(asset_type=AssetType.CONDITIONAL,
                           token_id=token_id,
                           signature_type=SignatureTypeV2.POLY_1271)
)

# Check on-chain YES token balance
ctf_c    = w3.eth.contract(address=CTF, abi=ERC1155_ABI)
yes_raw  = ctf_c.functions.balanceOf(DW, int(token_id)).call()
holdings = yes_raw / 1e6
if holdings < 0.01:
    print("⚠ No YES tokens in deposit wallet — run 7_market_buy_order.py first"); exit(1)

book = client.get_order_book(token_id)
bids = sorted(book["bids"], key=lambda b: float(b["price"]), reverse=True)
if not bids:
    print("No bids — cannot market sell"); exit(1)

# 1 tick below best bid to guarantee fill
price = round(float(bids[0]["price"]) - tick, 4)
size  = fok_size(holdings, price)

if size <= 0:
    print(f"⚠ No valid size for price={price} and holdings={holdings}"); exit(1)

print(f"YES holdings : {holdings:.5f} tokens")
print(f"Valid size   : {size}  (step constraint for price {price})")
print(f"Placing FOK market SELL: {size} shares @ ${price}  (best bid: ${float(bids[0]['price'])})")

resp = client.create_and_post_order(
    OrderArgs(token_id=token_id, price=price, size=size, side=Side.SELL),
    PartialCreateOrderOptions(tick_size=tick_size, neg_risk=neg_risk),
    OrderType.FOK,
)
order_id = resp.get("orderID") if isinstance(resp, dict) else resp.orderID
status   = resp.get("status")  if isinstance(resp, dict) else resp.status
print(f"\n✓ Order result : {order_id}  status: {status}")
